### ISSUE TITLE
loss function for D in wgan2-keras.py

### ISSUE THREAD

**Issue by jiangzidong** [tjwei/GANotebooks#1]
Hi

I think the loss function in wgan2-keras should be 
```
loss = loss_fake - loss_real + nz * λ * grad_penalty
```
in your In[9]

Because you called the `loss_mixed = K.mean(netD(netD_mixed_input))` before, which divide `nz`, so you should make up here.

Am I right?

**Comment by tjwei** [comment]
The output of `netD` has shape `(None, 1)`, so the mean is taken over `batchSize` items. 
That is, `loss_mixed` is a number, just like `loss_real` and `loss_fake`.
The shape of `grad_mixed = K.gradients(loss_mixed, [netD_mixed_input])[0]` is `(nc, imageSize, imageSize)`

This is what I meant in the code and hopefully the same as the original paper says.

**Comment by jiangzidong** [comment]
![image](https://user-images.githubusercontent.com/15059452/27731758-f3b08c90-5dc0-11e7-82b8-9ed532ddc3b8.png)

line 7: the L is for each samples in the batch
line 9, get the batch mean of the L. You have `grad_penalty = K.mean(K.square(norm_grad_mixed -1))`, `loss_real = K.mean(netD(netD_real_input))` and `loss_fake = K.mean(netD(netD_fake_input))`, which corresponds the 3 items in L.

But before that, you also called `loss_mixed = K.mean(netD(netD_mixed_input))`, but in the original paper, there is no other mean operations. 

For example, batch size is 2, we have 2 mixed input `x1`, and `x2`, and their netD output is `o1`, and `o2`, and their gradients are `g1 = delta_o1/delta_x1`, `g2 = delta_o2/delta_x2`, then we do the calculation in 7 & 9.
But in your code, you firstly get the means of `o1, o2`: `o_mean = (o1+o2)/2`, and then the gradient became `g1_now = 0.5* delta_o1/delta_x1 = 0.5*g1` (x1 has no relation with o2, and x2 has no relation with o1)

I am not sure if I put it clear...And I am not 100% sure I am right.

**Comment by tjwei** [comment]
Thanks. I think you are right. I will fix and rerun the experiments.

**Comment by tjwei** [comment]
Hopefully this commit got the formula  correct. The experiments are running, seems good so far.
https://github.com/tjwei/GANotebooks/commit/7424073f5962e99b46dcb69282c6eeeb8c77d5e5
